clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RequestBasedMetadata.java

```diff
@@ -502,6 +518,83 @@ synchronized void updateCache(boolean onDemandRefresh) throws InterruptedExcepti
     }
   }
 
+  public static boolean whetherToSwitchToFetchedCurrentVersion(
+      String storeName,
+      Set<Integer> activeVersions,
+      AtomicInteger currentVersion,
+      int fetchedCurrentVersion,
+      int partitionCountForFetchedCurrentVersion,
+      Map<Integer, List<String>> routingInfo) {
+    if (fetchedCurrentVersion == currentVersion.get()) {
+      // Already same
+      return false;
+    }
+
+    boolean isPartitionResourcesReadyForFetchedCurrentVersion = isPartitionResourcesReady(
+        storeName,
+        fetchedCurrentVersion,
+        partitionCountForFetchedCurrentVersion,
+        routingInfo);
+    boolean isCurrentVersionActive = activeVersions.contains(currentVersion.get());
+    if (!isCurrentVersionActive || isPartitionResourcesReadyForFetchedCurrentVersion) {
+      if (!isCurrentVersionActive && !isPartitionResourcesReadyForFetchedCurrentVersion) {
+        LOGGER.warn(
+            "Current version: {} is not active anymore for store: {}, updating to fetched current version: {}"
+                + " even the partition resources are not ready",
+            currentVersion.get(),
+            storeName,
+            fetchedCurrentVersion);
+      }
+      return true;
+    }
+    // If the partition resources are not ready for the fetched current version, we do not update the current
+    // version to avoid breaking the client.
+    LOGGER.warn(
```

Comment:
Similar here, should we do info instead of warn? Or upgrade to warn once we reach certain timeout/attempt. e.g. 5 times/minutes.